src/spyglass/utils/dj_mixin.py

```diff
@@ -535,3 +536,91 @@ def super_delete(self, *args, **kwargs):
         logger.warning("!! Using super_delete. Bypassing cautious_delete !!")
         self._log_use(start=time(), super_delete=True)
         super().delete(*args, **kwargs)
+
+    from spyglass.utils.dj_merge_tables import Merge
+
+    def restrict_from_upstream(self, key, **kwargs):
+        """Recursive function to restrict a table based on secondary keys of upstream tables"""
+        return restrict_from_upstream(self, key, **kwargs)
+
+
+def restrict_from_upstream(table, key, max_recursion=3):
```

Comment:
I'm thinking I could fold this into either `dj_chains` or `dj_graph` to make use of how restr are cascaded there. it might cutdown on the footprint for this feature